source/_posts/2024-06-07-ai-agents-for-the-smart-home.markdown

```diff
@@ -0,0 +1,189 @@
+
+We've used these tools extensively to fine tune the prompt and API that we give to LLMs to control Home Assistant. The reproducibility of these studies allows us to change something and repeat the test to see if we can generate better results. We are able to use this to test different prompts, different AI models and any other aspect.
+
+## Defining the API for LLMs
+
+Home Assistant has different API interfaces. We have the Home Assistant Python object, a WebSocket API, Rest API, and intents. We decided to base our LLM API on the intent system because it is our smallest API. Intents are used by our sentence-matching voice assistant and are limited to controlling devices and querying information. They don't bother with creating automations, managing devices, and other administrative tasks.
+
+Leveraging intents also meant that we already have a place in the UI where users can configure what entities are accessible, a test suite in many languages matching sentences to intent and a baseline of what the LLM should be able to achieve with the API.
+
+<img src='/images/blog/2024-06-ai-agents-for-the-smart-home/06-expose.png' alt='Home Assistant Conference header' class='no-shadow'>
+
+Home Assistant already has different ways for a user to define their own intents, and this allows users to extend the Assist API that LLMs have access to. The first one is the [intent script integration](/integrations/intent_script/). Using YAML, users can define a script to run when the intent is invoked and use a template to define the response.
+
+{% highlight python %}
+{% raw %}
+intent_script:
+  EventCountToday:
+    action:
+      - service: calendar.get_events
+        target:
+          entity_id: calendar.my_calendar
+        data_template:
+          start_date_time: "{{ today_at('00:00') }}"
+          duration: { "hours": 24 }
+        response_variable: result
+      - stop: ""
+        response_variable: result
+    speech:
+      text: "{{ action_response['calendar.my_calendar'].events | length }} events"
+{% endraw %}
+{% endhighlight %}
+
+We haven't forgotten about custom components either. They can [register their own intents](https://developers.home-assistant.io/docs/intent_handling) or even better: define their own API.
+
+### Custom integrations providing their own LLM APIs
+
+The built-in LLM API is focused on simplicity and being good at the things that it does. The larger the API surface, the easier AI models, especially the smaller ones, can get confused and invoke them incorrectly.
+
+So instead of one large API, we are aiming for many focused APIs. An AI agent will only have access to one API at a time to ensure it has a higher success rate. It will require experimentation to figure out the best API for creating automations, querying the history, and maybe even creating dashboards. When all those APIs are in place, we can start playing with a selector agent that routes incoming requests to the right agent and API.
```

Comment:
```suggestion
Instead of one large API, we are aiming for many focused APIs. To ensure a higher success rate, an AI agent will only have access to one API at a time. Figuring out the best API for creating automations, querying the history, and maybe even creating dashboards will require experimentation. When all those APIs are in place, we can start playing with a selector agent that routes incoming requests to the right agent and API.
```